# <center>MDS6212 Fintech Theory and Practice: Week 1 Assignment<center>

----
    
**220041006 *Yihang Li***

**Data confidential requirements:** 
*     Do not share datasets with others
*     Do not post datasets on the Web site
*     Datasets can only be used to work on assignments of MDS6212, M.Sc. in Data Science, CUHK-Shenzhen.

**Datasets:** 
*     Week 1 Data.csv
*     Week 1 Variable descriptions.xlsx

> Reference, Dai, Lili, Jianlei Han, Jing Shi, and Bohui Zhang, 2020, Digital Footprints as Collateral for Debt Collection, working paper.

----

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Q1" data-toc-modified-id="Q1-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Q1</a></span><ul class="toc-item"><li><span><a href="#Description-for-numeric-variables" data-toc-modified-id="Description-for-numeric-variables-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Description for numeric variables</a></span></li><li><span><a href="#Description-for-categorical-variables" data-toc-modified-id="Description-for-categorical-variables-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Description for categorical variables</a></span></li><li><span><a href="#Dealing-with-missing-values:-'nominalrates':-3,-'creditlevelasbuyer':-969,-'default':-2795" data-toc-modified-id="Dealing-with-missing-values:-'nominalrates':-3,-'creditlevelasbuyer':-969,-'default':-2795-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Dealing with missing values: 'nominalrates': 3, 'creditlevelasbuyer': 969, 'default': 2795</a></span><ul class="toc-item"><li><span><a href="#nominalrates-and-creditlevelasbuyer" data-toc-modified-id="nominalrates-and-creditlevelasbuyer-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>nominalrates and creditlevelasbuyer</a></span></li><li><span><a href="#default" data-toc-modified-id="default-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>default</a></span></li></ul></li></ul></li><li><span><a href="#Q2" data-toc-modified-id="Q2-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Q2</a></span><ul class="toc-item"><li><span><a href="#Using-LogisticRegression-from-sklearn" data-toc-modified-id="Using-LogisticRegression-from-sklearn-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Using LogisticRegression from sklearn</a></span></li><li><span><a href="#Using-Logit-from-statsmodels" data-toc-modified-id="Using-Logit-from-statsmodels-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Using Logit from statsmodels</a></span></li></ul></li><li><span><a href="#Q3" data-toc-modified-id="Q3-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Q3</a></span><ul class="toc-item"><li><span><a href="#Using-LogisticRegression-from-sklearn" data-toc-modified-id="Using-LogisticRegression-from-sklearn-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Using LogisticRegression from sklearn</a></span></li><li><span><a href="#Using-Logit-from-statsmodels" data-toc-modified-id="Using-Logit-from-statsmodels-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Using Logit from statsmodels</a></span></li></ul></li><li><span><a href="#Q4" data-toc-modified-id="Q4-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Q4</a></span><ul class="toc-item"><li><span><a href="#Using-LogisticRegression-from-sklearn" data-toc-modified-id="Using-LogisticRegression-from-sklearn-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Using LogisticRegression from sklearn</a></span></li><li><span><a href="#Using-Logit-from-statsmodels" data-toc-modified-id="Using-Logit-from-statsmodels-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Using Logit from statsmodels</a></span></li></ul></li></ul></div>

In [1]:
import pandas as pd
import numpy as np

In [2]:
Data = pd.read_csv('./MDS6212 Week 1 Data.csv', encoding='gbk', skipinitialspace=True)

In [3]:
Data.head()

uid          id  provincecode  birthday  age  gender  areaid  \
0   72146187  1200345737            43  19951203   24   False  431222   
1  101101513  1200456775            32  19880607   31   False  320382   
2   98608810  1200502203            42  19960503   23   False  420821   
3   96300664  1200424426            51  20000620   19    True  510824   
4  105585598  1200512684            32  19940122   25   False  321324   

  provincename cityname areaname  ...  highcontact  highcontact20s  \
0          湖南省      怀化市      沅陵县  ...        False           False   
1          江苏省      徐州市      邳州市  ...         True            True   
2          湖北省      荆门市      京山县  ...         True            True   
3          四川省      广元市      苍溪县  ...        False           False   
4          江苏省      宿迁市      泗洪县  ...         True            True   

  numbercontacttotal default_firstmonth_0 default_firstmonth_30  \
0                  0                  NaN                   NaN   
1                 47                False                 False   
2                 23                  NaN                   NaN   
3                  0                False                 False   
4                 46                False                 False   

  default_firstmonth_60  delaydate_max  default_max_0  default_max_30  \
0                   NaN            NaN            NaN             NaN   
1                 False           -3.0          False           False   
2                   NaN            NaN            NaN             NaN   
3                 False          -23.0          False           False   
4                 False            0.0          False           False   

   default_max_60  
0             NaN  
1           False  
2             NaN  
3           False  
4           False  

[5 rows x 47 columns]

## Q1

*1) Present a table of summary statistics for the key variables including the borrower’s age, gender, loan amount, interest rate, credit scores, a dummy whether the borrower has a frequent contact, approval dummy, and delinquency dummy*

In [4]:
Data.columns

Index(['uid', 'id', 'provincecode', 'birthday', 'age', 'gender', 'areaid',
       'provincename', 'cityname', 'areaname', 'instalments_num',
       'instalments_amount', 'shop_address', 'trans_date', 'taobaostartdate',
       'taobaoenddate', 'taobaodealno', 'apptimes', 'deal',
       'apply_request_sum', 'apply_reject_sum', 'loan_offer_sum',
       'repay_fail_sum', 'max_default_days', 'default', 'province',
       'nominalrates', 'default30', 'alipaybalance', 'yuebaobalance',
       'huabeiamount', 'huabeibalance', 'creditlevelasbuyer', 'tencentscore',
       'gaodescore', 'numbercontact', 'numbercontact20s', 'highcontact',
       'highcontact20s', 'numbercontacttotal', 'default_firstmonth_0',
       'default_firstmonth_30', 'default_firstmonth_60', 'delaydate_max',
       'default_max_0', 'default_max_30', 'default_max_60'],
      dtype='object')

> Note: instalments_amount is the loan_amount

In [5]:
#Collect the numeric variables and categorical variables mentioned in Q1
Key_num_Var_list = [
    'age',
    'instalments_amount',
    'nominalrates',
    'creditlevelasbuyer',
    'tencentscore',
    'gaodescore',
]
Key_cat_Var_list = ['gender', 'highcontact', 'deal', 'default']

### Description for numeric variables

In [6]:
Data[Key_num_Var_list].head()

age  instalments_amount  nominalrates  creditlevelasbuyer  tencentscore  \
0   24              603000      0.494169                 NaN        60.200   
1   31              270000      0.359378                 1.0        48.000   
2   23              398000      0.204553                 3.0        60.200   
3   19              300000      0.204560               161.0        53.875   
4   25              378000      0.252571                29.0        35.000   

   gaodescore  
0    0.192094  
1    0.192094  
2    0.156100  
3    0.192094  
4    0.189800

In [7]:
Data[Key_num_Var_list].describe().apply(
    lambda s: s.apply(lambda x: format(x, 'f')))

age instalments_amount nominalrates creditlevelasbuyer  \
count  5000.000000        5000.000000  4997.000000        4031.000000   
mean     27.675400      406201.420000     0.276058          53.119077   
std       8.326146      130623.360240     0.085912         108.629757   
min      18.000000       50000.000000     0.130080           0.000000   
25%      21.000000      320000.000000     0.204560           0.000000   
50%      25.000000      398000.000000     0.204579          14.000000   
75%      32.000000      498000.000000     0.359347          58.000000   
max      56.000000      869000.000000     0.494185        1830.000000   

      tencentscore   gaodescore  
count  5000.000000  5000.000000  
mean     58.608168     0.201975  
std      14.218112     0.076724  
min       9.000000     0.023518  
25%      53.888889     0.192094  
50%      60.200000     0.192094  
75%      65.258929     0.192094  
max      98.000000     0.732120

In [8]:
print(Data[['nominalrates']][pd.isnull(Data[['nominalrates']]).any(axis=1)])

      nominalrates
266            NaN
1452           NaN
4184           NaN


In [9]:
print(Data[['creditlevelasbuyer']][pd.isnull(Data[['creditlevelasbuyer']]).any(axis=1)])

      creditlevelasbuyer
0                    NaN
8                    NaN
14                   NaN
23                   NaN
27                   NaN
...                  ...
4976                 NaN
4982                 NaN
4984                 NaN
4985                 NaN
4998                 NaN

[969 rows x 1 columns]


*Number of missing values: 'nominalrates': 3, 'creditlevelasbuyer': 969*

### Description for categorical variables

In [10]:
Data[Key_cat_Var_list].head()

gender  highcontact  deal default
0   False        False     0     NaN
1   False         True     1   False
2   False         True     0     NaN
3    True        False     1   False
4   False         True     1   False

In [11]:
Data[Key_cat_Var_list].apply(lambda s: s.value_counts())
#For 'deal', True: 1, False: 0

gender  highcontact  deal  default
False    4267         2539  2793     1280
True      733         2461  2207      925

In [12]:
print(Data[['default']][pd.isnull(Data[['default']]).any(axis=1)])

     default
0        NaN
2        NaN
7        NaN
10       NaN
11       NaN
...      ...
4989     NaN
4990     NaN
4995     NaN
4997     NaN
4999     NaN

[2795 rows x 1 columns]


*Number of missing values: 'default': 2795*

In [13]:
#Not Neccessary
# #Convert True to 1, False to 0
# Data.gender = Data.gender.map({True:1, False:0})#.fillna(Data.gender)
# Data.highcontact = Data.highcontact.map({True:1, False:0})
# Data.default = Data.default.map({True:1, False:0})

### Dealing with missing values: 'nominalrates': 3, 'creditlevelasbuyer': 969, 'default': 2795

In [14]:
Key_Data = Data[Key_num_Var_list + Key_cat_Var_list]

In [15]:
Key_Data.head()

age  instalments_amount  nominalrates  creditlevelasbuyer  tencentscore  \
0   24              603000      0.494169                 NaN        60.200   
1   31              270000      0.359378                 1.0        48.000   
2   23              398000      0.204553                 3.0        60.200   
3   19              300000      0.204560               161.0        53.875   
4   25              378000      0.252571                29.0        35.000   

   gaodescore  gender  highcontact  deal default  
0    0.192094   False        False     0     NaN  
1    0.192094   False         True     1   False  
2    0.156100   False         True     0     NaN  
3    0.192094    True        False     1   False  
4    0.189800   False         True     1   False

#### nominalrates and creditlevelasbuyer

*For nominalrates, since there are only 3 missing values, we simply delete the row of those three.*

In [16]:
Key_Data = Key_Data[Key_Data['nominalrates'].notnull()]
Key_Data.reset_index(drop = True, inplace = True)

*For convinient, we also delete the rows where creditlevelasbuyer is missing, because its missing values is not so much.*

In [17]:
Key_Data = Key_Data[Key_Data['creditlevelasbuyer'].notnull()]
Key_Data.reset_index(drop = True, inplace = True)

In [18]:
Key_Data.describe()

age  instalments_amount  nominalrates  creditlevelasbuyer  \
count  4029.000000         4029.000000   4029.000000         4029.000000   
mean     27.384711       405224.447754      0.268659           52.924795   
std       8.143781       128196.165282      0.078455          108.302046   
min      18.000000        65000.000000      0.130080            0.000000   
25%      21.000000       320000.000000      0.204560            0.000000   
50%      25.000000       398000.000000      0.204578           14.000000   
75%      32.000000       490000.000000      0.359347           58.000000   
max      56.000000       869000.000000      0.494185         1830.000000   

       tencentscore   gaodescore         deal  
count   4029.000000  4029.000000  4029.000000  
mean      58.547673     0.202780     0.451229  
std       14.449751     0.074965     0.497677  
min       10.000000     0.023518     0.000000  
25%       53.000000     0.192094     0.000000  
50%       60.200000     0.192094     0.000000  
75%       66.000000     0.192094     1.000000  
max       98.000000     0.732120     1.000000

*By doing so, we can focus on dealing with the huge propotion missing values of default*

*Now we have 4029 observations in total*

#### default

*The idea is: Consider default vs other variables in the Key_Data, use non-missing data to build a logistic classifier, and then use such classifier to predict the value of missing default.*

In [19]:
nm_Data = Key_Data[Key_Data['default'].notnull()]
nm_Data.reset_index(drop = True, inplace = True)

In [20]:
nm_Data.describe()

age  instalments_amount  nominalrates  creditlevelasbuyer  \
count  1850.000000         1850.000000   1850.000000         1850.000000   
mean     26.816757       389656.216216      0.282377           68.390811   
std       7.473439       121658.903413      0.078242          125.680866   
min      18.000000        65000.000000      0.130080            0.000000   
25%      21.000000       300000.000000      0.204561            3.000000   
50%      25.000000       398000.000000      0.252573           26.000000   
75%      31.000000       438000.000000      0.359347           77.750000   
max      55.000000       869000.000000      0.494180         1830.000000   

       tencentscore   gaodescore         deal  
count   1850.000000  1850.000000  1850.000000  
mean      55.282119     0.193852     0.979459  
std       15.979269     0.070673     0.141878  
min       10.000000     0.023518     0.000000  
25%       44.607143     0.187434     1.000000  
50%       58.550000     0.192094     1.000000  
75%       66.160714     0.192094     1.000000  
max       96.166667     0.638378     1.000000

*we have 1850 non-missing values in total*

In [21]:
nm_Data.head(1)

age  instalments_amount  nominalrates  creditlevelasbuyer  tencentscore  \
0   31              270000      0.359378                 1.0          48.0   

   gaodescore  gender  highcontact  deal default  
0    0.192094   False         True     1   False

In [22]:
nm_X = nm_Data[[i for i in nm_Data.columns if i not in ['default']]]
nm_y = nm_Data['default']

In [23]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
#We can also use Statsmodels.api

In [24]:
scaler = StandardScaler()
nm_clf = LogisticRegression()
nm_X = scaler.fit(nm_X).transform(nm_X)

In [25]:
nm_clf.fit(nm_X, nm_y.astype(bool))

LogisticRegression()

In [26]:
m_Data = Key_Data[Key_Data['default'].isnull()]
m_Data.reset_index(drop = True, inplace = True)

In [27]:
m_X = m_Data[[i for i in nm_Data.columns if i not in ['default']]]
m_X = scaler.fit(m_X).transform(m_X)

In [28]:
nm_clf.predict(m_X)

array([False, False, False, ..., False,  True, False])

In [29]:
Key_Data['default'].iloc[Key_Data[Key_Data['default'].isnull()].index] = nm_clf.predict(m_X)

*Now we have filled the missing values of 'default'*

In [30]:
Key_Data['default'].isnull().any()

False

In [31]:
Key_Data

age  instalments_amount  nominalrates  creditlevelasbuyer  tencentscore  \
0      31              270000      0.359378                 1.0     48.000000   
1      23              398000      0.204553                 3.0     60.200000   
2      19              300000      0.204560               161.0     53.875000   
3      25              378000      0.252571                29.0     35.000000   
4      21              400000      0.359360                54.0     84.000000   
...   ...                 ...           ...                 ...           ...   
4024   33              397000      0.204564                 0.0     13.000000   
4025   32              458000      0.204559               146.0     53.000000   
4026   21              130000      0.358954                17.0     58.111111   
4027   24              589000      0.213772                11.0     70.571429   
4028   20              139000      0.359367                 9.0     60.200000   

      gaodescore  gender  highcontact  deal default  
0       0.192094   False         True     1   False  
1       0.156100   False         True     0   False  
2       0.192094    True        False     1   False  
3       0.189800   False         True     1   False  
4       0.187895   False        False     1   False  
...          ...     ...          ...   ...     ...  
4024    0.214644   False         True     1   False  
4025    0.305604   False         True     0   False  
4026    0.192094   False         True     1    True  
4027    0.241720   False         True     0    True  
4028    0.192094    True         True     0   False  

[4029 rows x 10 columns]

## Q2

*2) Perform a logit regression and examine the relation between the delinquency likelihood and credit scores*

### Using LogisticRegression from sklearn

In [32]:
q2_X = Key_Data[['creditlevelasbuyer', 'tencentscore', 'gaodescore']]
q2_y = Key_Data['default']
q2_X = scaler.fit(q2_X).transform(q2_X)

In [33]:
q2_logit = LogisticRegression().fit(q2_X, q2_y.astype(bool))

score:float
* Mean accuracy of self.predict(X) wrt. y.

In [34]:
q2_logit.score(q2_X, q2_y.astype(bool))

0.7217671878878134

In [35]:
q2_logit.coef_

array([[0.10413267, 0.31336157, 0.13033775]])

### Using Logit from statsmodels

In [36]:
import statsmodels.api as sm

In [37]:
q2_X = Key_Data[['creditlevelasbuyer', 'tencentscore', 'gaodescore']]
q2_y = Key_Data['default'].map({True:1, False:0})
q2_X = scaler.fit(q2_X).transform(q2_X)
q2_X = sm.add_constant(q2_X)

In [38]:
q2_model = sm.Logit(q2_y, q2_X)
q2_result = q2_model.fit()

Optimization terminated successfully.
         Current function value: 0.580266
         Iterations 5


In [39]:
q2_result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                default   No. Observations:                 4029
Model:                          Logit   Df Residuals:                     4025
Method:                           MLE   Df Model:                            3
Date:                Tue, 24 Nov 2020   Pseudo R-squ.:                 0.01901
Time:                        10:25:04   Log-Likelihood:                -2337.9
converged:                       True   LL-Null:                       -2383.2
Covariance Type:            nonrobust   LLR p-value:                 1.605e-19
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.9781      0.036    -27.192      0.000      -1.049      -0.908
x1             0.1044      0.035      2.979      0.003       0.036       0.173
x2             0.3138      0.038      8.161      0.000       0.238       0.389
x3             0.1305      0.035      3.765      0.000       0.063       0.198
==============================================================================
"""

In [41]:
#for creditlevelasbuyer
q21_X = Key_Data['creditlevelasbuyer']
q21_y = Key_Data['default'].map({True:1, False:0})
q21_X = scaler.fit(q21_X.values.reshape(-1, 1)).transform(q21_X.values.reshape(-1, 1))
q21_X = sm.add_constant(q21_X)
q21_model = sm.Logit(q21_y, q21_X)
q21_result = q21_model.fit()
q21_result.summary()

Optimization terminated successfully.
         Current function value: 0.591422
         Iterations 5


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                default   No. Observations:                 4029
Model:                          Logit   Df Residuals:                     4027
Method:                           MLE   Df Model:                            1
Date:                Tue, 24 Nov 2020   Pseudo R-squ.:               0.0001534
Time:                        10:30:18   Log-Likelihood:                -2382.8
converged:                       True   LL-Null:                       -2383.2
Covariance Type:            nonrobust   LLR p-value:                    0.3925
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.9522      0.035    -27.088      0.000      -1.021      -0.883
x1             0.0294      0.034      0.863      0.388      -0.037       0.096
==============================================================================
"""

In [42]:
q22_X = Key_Data['tencentscore']
q22_y = Key_Data['default'].map({True:1, False:0})
q22_X = scaler.fit(q22_X.values.reshape(-1, 1)).transform(q22_X.values.reshape(-1, 1))
q22_X = sm.add_constant(q22_X)
q22_model = sm.Logit(q22_y, q22_X)
q22_result = q22_model.fit()
q22_result.summary()

Optimization terminated successfully.
         Current function value: 0.582737
         Iterations 5


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                default   No. Observations:                 4029
Model:                          Logit   Df Residuals:                     4027
Method:                           MLE   Df Model:                            1
Date:                Tue, 24 Nov 2020   Pseudo R-squ.:                 0.01484
Time:                        10:47:25   Log-Likelihood:                -2347.8
converged:                       True   LL-Null:                       -2383.2
Covariance Type:            nonrobust   LLR p-value:                 4.125e-17
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.9730      0.036    -27.169      0.000      -1.043      -0.903
x1             0.3077      0.038      8.181      0.000       0.234       0.381
==============================================================================
"""

In [43]:
#q2_X = Key_Data[['creditlevelasbuyer', 'tencentscore', 'gaodescore']]
q22_X = Key_Data['gaodescore']
q22_y = Key_Data['default'].map({True:1, False:0})
q22_X = scaler.fit(q22_X.values.reshape(-1, 1)).transform(q22_X.values.reshape(-1, 1))
q22_X = sm.add_constant(q22_X)
q22_model = sm.Logit(q22_y, q22_X)
q22_result = q22_model.fit()
q22_result.summary()

Optimization terminated successfully.
         Current function value: 0.589258
         Iterations 5


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                default   No. Observations:                 4029
Model:                          Logit   Df Residuals:                     4027
Method:                           MLE   Df Model:                            1
Date:                Tue, 24 Nov 2020   Pseudo R-squ.:                0.003813
Time:                        10:50:44   Log-Likelihood:                -2374.1
converged:                       True   LL-Null:                       -2383.2
Covariance Type:            nonrobust   LLR p-value:                 2.018e-05
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.9564      0.035    -27.107      0.000      -1.026      -0.887
x1             0.1454      0.034      4.303      0.000       0.079       0.212
==============================================================================
"""

## Q3

*3) Perform a logit regression and examine the relation between the loan approval likelihood and credit scores*

### Using LogisticRegression from sklearn

In [40]:
q3_X = Key_Data[['creditlevelasbuyer', 'tencentscore', 'gaodescore']]
q3_y = Key_Data['deal']
q3_X = scaler.fit(q3_X).transform(q3_X)

In [41]:
q3_logit = LogisticRegression().fit(q3_X, q3_y.astype(bool))

In [42]:
q3_logit.score(q3_X, q3_y.astype(bool))

0.6443286175229586

In [43]:
q3_logit.coef_

array([[ 0.20166668, -0.39714435, -0.15879942]])

### Using Logit from statsmodels

In [44]:
q3_X = Key_Data[['creditlevelasbuyer', 'tencentscore', 'gaodescore']]
q3_y = Key_Data['deal']
q3_X = scaler.fit(q3_X).transform(q3_X)
q3_X = sm.add_constant(q3_X)

In [45]:
q3_model = sm.Logit(q3_y, q3_X)
q3_result = q3_model.fit()

Optimization terminated successfully.
         Current function value: 0.658019
         Iterations 5


In [46]:
q3_result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                   deal   No. Observations:                 4029
Model:                          Logit   Df Residuals:                     4025
Method:                           MLE   Df Model:                            3
Date:                Sat, 26 Sep 2020   Pseudo R-squ.:                 0.04411
Time:                        20:28:39   Log-Likelihood:                -2651.2
converged:                       True   LL-Null:                       -2773.5
Covariance Type:            nonrobust   LLR p-value:                 9.324e-53
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.1993      0.033     -6.098      0.000      -0.263      -0.135
x1             0.2019      0.039      5.182      0.000       0.126       0.278
x2            -0.3976      0.035    -11.445      0.000      -0.466      -0.329
x3            -0.1589      0.035     -4.580      0.000      -0.227      -0.091
==============================================================================
"""

## Q4

*4) Perform a logit regression and examine the relation between the loan approval likelihood and the dummy whether the borrower has a frequent contact*

### Using LogisticRegression from sklearn


In [47]:
q4_X = Key_Data['highcontact']
q4_y = Key_Data['deal']
q4_X = scaler.fit(q4_X.values.reshape(-1, 1)).transform(q4_X.values.reshape(-1, 1))

In [48]:
q4_logit = LogisticRegression().fit(q4_X, q4_y.astype(bool))

In [49]:
q4_logit.score(q4_X, q4_y.astype(bool))

0.5487714072970961

In [50]:
q4_logit.coef_

array([[0.08461788]])

### Using Logit from statsmodels

In [51]:
q4_X = Key_Data['highcontact']
q4_y = Key_Data['deal']
q4_X = scaler.fit(q4_X.values.reshape(-1, 1)).transform(q4_X.values.reshape(-1, 1))
q4_X = sm.add_constant(q4_X)

In [52]:
q4_model = sm.Logit(q4_y, q4_X)
q4_result = q4_model.fit()

Optimization terminated successfully.
         Current function value: 0.687495
         Iterations 4


In [53]:
q4_result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                   deal   No. Observations:                 4029
Model:                          Logit   Df Residuals:                     4027
Method:                           MLE   Df Model:                            1
Date:                Sat, 26 Sep 2020   Pseudo R-squ.:                0.001289
Time:                        20:28:39   Log-Likelihood:                -2769.9
converged:                       True   LL-Null:                       -2773.5
Covariance Type:            nonrobust   LLR p-value:                  0.007503
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.1961      0.032     -6.187      0.000      -0.258      -0.134
x1             0.0847      0.032      2.672      0.008       0.023       0.147
==============================================================================
"""